src/shims/unix/fd.rs

```diff
@@ -41,24 +43,19 @@ pub trait FileDescriptor: std::fmt::Debug + Any {
         throw_unsup_format!("cannot seek on {}", self.name());
     }
 
-    fn close<'tcx>(
-        self: Box<Self>,
-        _communicate_allowed: bool,
-    ) -> InterpResult<'tcx, io::Result<i32>> {
+    /// Note: This function should not be called except [`FileDescriptor::close`].
+    fn close<'tcx>(&mut self, _communicate_allowed: bool) -> InterpResult<'tcx, io::Result<i32>> {
```

Comment:
Can you make this function take `self`? Then we can be sure it is called only once.

We can use `Rc::into_inner` to get ownership of `self` for this.

Reply:
That's correct, using `self` as a parameter for `close` would be more natural. However, it would result in an error:

>  the size for values of type `Self` cannot be known at compilation time

An alternative approach is to take `Box<self>` as parameter and `FileDescriptor` will become `Rc<RefCell<Box<dyn FileDescription>>>`. What do you think about this?

Reply:
Ah, right, it must be object safe... the double pointer indirection is a bit unfortunate,y but yeah making it `Box<self>` sounds like a good idea to me.

Reply:
done :)